src/translate/invariants.py

```diff
@@ -77,125 +83,194 @@ def ensure_conjunction_sat(system, *parts):
+            other_arg_to_pos[arg].append(pos)
+
         factored_mapping = []
 
+        # We iterate over all values occuring as arguments in other_literal
+        # and compare the number of occurrences in other_literal to those in
+        # own_literal. If the difference of these numbers allows us to cover
+        # other_literal with the (still) permitted number of counted variables,
+        # we store the correspondance of all argument positions of
+        # other_literal for this value to the invariant parameters at these
+        # positions in factored_mapping. If all values can be covered, we
+        # instatiate the complete factored_mapping, computing all possibilities
+        # to map positions from other_literal to invariant parameters (or -1 if
+        # the position is omitted).
         for key, other_positions in other_arg_to_pos.items():
-            own_positions = arg_to_ordered_pos.get(key, [])
-            len_diff = len(own_positions) - len(other_positions)
+            inv_params = ownarg_to_invariant_parameters[key]
+            # all positions at which key occurs as argument in own_literal
+            len_diff = len(inv_params) - len(other_positions)
             if len_diff >= 1 or len_diff <= -2 or len_diff == -1 and not allowed_omissions:
+                # mapping of the literals is not possible with at most one
+                # counted variable.
                 return []
             if len_diff:
-                own_positions.append(-1)
+                inv_params.append(-1)
                 allowed_omissions = 0
-            factored_mapping.append((other_positions, own_positions))
+            factored_mapping.append((other_positions, inv_params))
         return instantiate_factored_mapping(factored_mapping)
 
     def possible_matches(self, own_literal, other_literal):
+        """This method is used when an action had an unbalanced add effect
+           on own_literal. The action has a delete effect on literal
+           other_literal, so we try to refine the invariant such that it also
+           covers the delete effect.
+
+           For this purpose, we consider all possible mappings from the
+           parameter positions of other_literal to the parameter positions of
+           own_literal such that the extended invariant can use other_literal
+           to balance own_literal. From these position mapping, we can extract
+           the new invariant part.
+
+           Consider for an example of the "self" InvariantPart "forall ?@v0,
+           ?@v1, ?@v2 P(?@v0, ?@v1, ?@v2) is non-increasing" and let
+           own_literal be P(?a, ?b, ?c) and other_literal be Q(?b, ?c, ?d, ?a).
+           The only possible mapping from positions of Q to invariant variables
+           of P (or -1) is [0->?@v1, 1->?@v2, 2->-1, 3->?@v0] for which we
+           create a new Invariant Part Q(?@v1, ?@v2, _. ?@v0) with the third
+           argument being counted.
```

Comment:
The extended version of the documentation is in the documentation of possible_mappings. Long story short: We create all three matchings as new invariant candidates.

Independent of this: I just noticed that the possible_matches could be more nicely implemented as a generator (yielding InvariantParts instead of first collecting them in a list). Would you also find this preferable?

Reply:
Ok good then I understood correctly what the code actually does. (I don't understand why this is what we want but as I mentioned before I didn't go into the details on what the entire invariant synthesis does, and instead just tried to understand what this piece of code does.)

I don't have a strong opinion on the generator approach but since quite some parts of the code work with generators it probably would be more consistent to do it wherever it makes sense.

Reply:
I went for the generator.